Adds a function to construct a Glushkov NFA from a Thompson NFA

sfstewman opened the pull request:
This ended up reusing almost all of the `fsm_determinise()` code, but required a different `buildtransitions`.

The names need a bit of an overhaul, and it could use additional fuzzing / testing.

katef commented on src/libfsm/determinise.c:
```diff
@@ -292,6 +285,129 @@ hash_single_state(const void *a)
+	assert(curr->closure != NULL);
+
+	sym_min = UCHAR_MAX+1;
+	sym_max = -1;
```
Rather than `UCHAR_MAX+1`, better to say `FSM_SIGMA_MAX`, if you're after an out-of-bounds value one past the maximum symbol in the alphabet.

sfstewman replied:
done: a32ef76

katef commented on src/libfsm/determinise.c:
```diff
@@ -292,6 +285,129 @@ hash_single_state(const void *a)
+	/* double check that the edge symbols are still within 0..UCHAR_MAX */
+	if (sym_min < 0 || sym_max > UCHAR_MAX) {
+		goto finish;
+	}
```
Likewise `FSM_SIGMA_MAX` here.

sfstewman replied:
done: a32ef76

katef commented on src/libfsm/determinise.c:
```diff
@@ -516,12 +632,19 @@ determinise(struct fsm *nfa,
 	for (curr = mapping_set_first(mappings, &it); (curr = nextnotdone(mappings, &sv)) != NULL; curr->done = 1) {
-		if (!buildtransitions(nfa, dfa, mappings, curr)) {
-			goto error;
+		/* XXX - may want to hoist this outside the loop to avoid branching */
```
Probably not, our bottlenecks are very much elsewhere

sfstewman replied:
removed comment: 0e11a68

katef commented on src/libfsm/determinise.c:
```diff
@@ -618,3 +749,18 @@ fsm_determinise(struct fsm *fsm)
+
+	return r;
+}
+
```
Would prefer to move out the guts to its own file, and have both the `fsm_determinise*()` and `fsm_glushkovize*()` interfaces in fsm_determinise.c and fsm_glushkovize.c respectively, calling into it. That'll help make naming clearer too, I hope.

But it'll make for a messy diff, so maybe on a separate PR?

sfstewman replied:
yes, I think a separate PR would be best.

katef commented on tests/glush/in1.fsm:
```diff
@@ -0,0 +1,19 @@
+#
+# Copyright 2008-2017 Katherine Flavel
```
You misspelled your name :)

sfstewman replied:
done: 0e11a68

sfstewman commented on src/libfsm/determinise.c:
```diff
@@ -292,6 +285,129 @@ hash_single_state(const void *a)
+
+finish:
+
+	for (sym = 0; sym < UCHAR_MAX+1; sym++) {
```
done: a32ef76

sfstewman commented on src/libfsm/determinise.c:
```diff
@@ -292,6 +285,129 @@ hash_single_state(const void *a)
+	for (sym = sym_min; sym <= sym_max; sym++) {
+		size_t i,n;
+
+		/* is there a way to bulk add these? */
```
done: 62c5edd and 15c1295

sfstewman commented:
Also:
* renamed `fsm_glushkovise()` in c04ed60
* Fixed comment in `fsm.h` in ed40603
* Added "documentation" or something that vaguely resembles it in 8217f6e
